### PR TITLE
.initial commit for .relocatable. python3 in basement

### PR DESCRIPTION
**Pull request by daxcore**
hey. 

this is a prototype of my idea to add python3 to basement: https://github.com/BobBuildTool/basement/issues/40

- relocatable, not 100%, but it does, what it shall
- some other necessary fixes
- a new class for python3 stuff, like a pip install
- added the usage-example to the existing libs: libholler and libgreeter.

surely not perfect, but maybe with power of the community a cool, new feature?!

### CODE REVIEW

File: recipes/python/python3.yaml [hunk 1]
```diff
@@ -53,27 +55,40 @@ multiPackage:
...
                 --disable-test-modules \
                 --disable-lib2to3 \
                 --disable-tk \
                 --disable-nis \
                 --disable-idle3 \
-                --disable-pyc-build
+                --disable-bzip2 \
+                --disable-xz \
+                --disable-pyc-build \
+                --disable-uuid
```
Comment (daxcore) [inline recipes/python/python3.yaml:70]: or we add e2fsprogs library... if necessary.

File: recipes/python/python3.yaml [hunk 2]
```diff
@@ -53,27 +55,40 @@ multiPackage:
...
 
+        provideTools:
+            python3: usr/bin
+
     minimal:
         depends:
            - libs::expat-dev
+           - libs::gdbm-dev
            - libs::libffi-dev
+           - libs::zlib-dev
```
Comment (daxcore) [inline recipes/python/python3.yaml:91]: still needed for gdbm (also if --disable-zlib), which is also a must-have...

File: tests/cross-platform/recipes/cmake/libholler.yaml [hunk 3]
```diff
@@ -7,11 +7,13 @@ checkoutSCM:
...
 
 multiPackage:
     dev:
         packageScript: |
            cmakePackageDev
+           installPythonModule
     tgt:
         packageScript: |
-           cmakePackageTgt
+           cmakePackageTgt "$1" "!/usr/lib/python*"
```
Comment (daxcore) [inline tests/cross-platform/recipes/cmake/libholler.yaml:19]: or we have to "fix" the install class. (full rebuild necessary)

File: tests/cross-platform/recipes/cmake/libgreet.yaml [hunk 4]
```diff
@@ -12,6 +12,23 @@ checkoutSCM:
     prune: True
 
 buildScript: |
+   cat <<EOF > myfile.py
```
Comment (daxcore) [inline tests/cross-platform/recipes/cmake/libgreet.yaml:15]: this could be e.g. a python-code-generator...

File: classes/python3.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+buildTools: [python3]
+buildVars: [PYTHONPATH]
+buildScript: |
+    #if [[ ${BOB_HOST_PLATFORM} == "linux" ]]
+    export PYTHONPATH
+    for i in "${!BOB_DEP_PATHS[@]}" ; do
+        if [[ $i == *-dev ]] ; then
+            for j in $(find ${BOB_DEP_PATHS[$i]}/usr -type d -name "python3*") ; do
```
Comment (daxcore) [inline classes/python3.yaml:8]: we can also get the paths by using pythons sysconfig module!

**Comment by jkloetzke** [comment]
Having a relocatable python as tool is definitely worth it. A couple of questions come to my mind:

* Wouldn't that replace `python3-minimal`? This is the current version of a very minimal python3 tool. If we can have a full featured, relocatable one then let's use this. The standard python3 package 
* Does this all still build in the sandbox? The ncurses `--enable-widec` and python3 `--with-ensurepip` look scary.
* The tests should be separate from the cmake tests. It also does not look cross-platform yet. Does this work on MSYS?

**Comment by daxcore** [comment]
in all my tests i needed `--enable-widec`, without it, python build will fail, while linking time:
```
cannot find -lpanelw
cannot find -lncursesw
```

`--with-ensurepip` is one of the main features to install the necessary modules.
but we could use the `get-pip.py` script instead. so we can set a fixed version number (no internal automatic version-changing: necessary for artifacts).

For MSYS2 we have to invest some more, but it should work. we have to use the "embed" version of python3 for windows.
We should not build it with the host-MSYS toolchain. We need a native windows python3 tool, or we will run into trouble with path-stuff later. (The same like, we have to use a native CMake tool, and can't use the MSYS2 CMake tool.)

another thing is the venv module of python3. but i got no good solution to use this instead of the `--root` `--prefix` of `python3 -m pip install`.

**Comment by daxcore** [comment]
> `--with-ensurepip` is one of the main features to install the necessary modules.
> but we could use the `get-pip.py` script instead. so we can set a fixed version number (no internal automatic version-changing: necessary for artifacts).

ohh fail, i see there is already a python3-setuptools recipe... so just adding pip and wheel in the same way and will not need all this stuff above.

**Comment by daxcore** [comment]
update: adapted to MSYS2 (including retest in linux) - works fine for me.

btw: will that be a problem, if python bytecode will be created in package-directories, when the packages will be imported? should we force `pyton3 -B` option?

File: classes/basement/rootrecipe.yaml [hunk 6]
```diff
@@ -65,6 +65,15 @@ depends:
         - name: utils::e2fsprogs
           use: [tools]
           forward: True
+        - name: python::python3-minimal
```
Comment (daxcore) [inline classes/basement/rootrecipe.yaml:68]: minimal: just python
"": python + setuptools
full: python + setuptools + pip + wheel

File: classes/python3.yaml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    done
+    export PYTHONPATH
+
+    pip_install()
+    {
+        # force lib instead of Lib for MSYS
+        mkdir -p install/usr/lib
+        python3 -m pip install $* \
+            --root ${PWD}/install \
+            --prefix usr \
```
Comment (daxcore) [inline classes/python3.yaml:35]: do not use "/usr". this will fail in MSYS, but just "usr" will also work in linux!

File: classes/python3.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            --isolated \
```
Comment (daxcore) [inline classes/python3.yaml:36]: maybe not necessary?!

File: tests/cross-platform/recipes/cmake/greeter.yaml [hunk 9]
```diff
@@ -18,11 +21,21 @@ packageScript: |
 
 multiPackage:
     host:
+        environment:
+            HOST: "1"
         depends: &deps
+            - if: !expr |
+                "${HOST}" == "1"
```
Comment (daxcore) [inline tests/cross-platform/recipes/cmake/greeter.yaml:28]: how to do that better? xD

**Comment by daxcore** [comment]
added patch for #44

**Comment by daxcore** [comment]
finished. works for me fine. what do u think about this?

**Comment by daxcore** [comment]
so this merge-request is obsolete now!
i restructured all changes in multiple merge-request, with clean commit-messages ;-)

I guess we should close this one here.

**Pull request closed without merging**